typing/env.mli

```diff
@@ -90,52 +95,130 @@ val add_required_global: Ident.t -> unit
+  | Cannot_scrape_alias of Longident.t * Path.t
+
+val lookup_error: Location.t -> t -> lookup_error -> 'a
+
 val lookup_constructor:
-  ?loc:Location.t -> ?mark:bool -> Longident.t -> t -> constructor_description
+  ?use:bool -> loc:Location.t -> constructor_usage -> Longident.t -> t ->
+  constructor_description
 val lookup_all_constructors:
-  ?loc:Location.t -> ?mark:bool ->
-  Longident.t -> t -> (constructor_description * (unit -> unit)) list
+  ?use:bool -> loc:Location.t -> constructor_usage -> Longident.t -> t ->
+  ((constructor_description * (unit -> unit)) list,
+   Location.t * t * lookup_error) result
+val lookup_all_constructors_from_type:
+  ?use:bool -> loc:Location.t -> constructor_usage -> Path.t -> t ->
+  (constructor_description * (unit -> unit)) list
+
 val lookup_label:
-  ?loc:Location.t -> ?mark:bool ->
-  Longident.t -> t -> label_description
+  ?use:bool -> loc:Location.t -> Longident.t -> t ->
+  label_description
 val lookup_all_labels:
-  ?loc:Location.t -> ?mark:bool ->
-  Longident.t -> t -> (label_description * (unit -> unit)) list
-val lookup_type:
-  ?loc:Location.t -> ?mark:bool -> Longident.t -> t -> Path.t
-  (* Since 4.04, this function no longer returns [type_description].
-     To obtain it, you should either call [Env.find_type], or replace
-     it by [Typetexp.find_type] *)
-val lookup_module:
-  load:bool -> ?loc:Location.t -> ?mark:bool -> Longident.t -> t -> Path.t
-val lookup_modtype:
-  ?loc:Location.t -> ?mark:bool ->
+  ?use:bool -> loc:Location.t -> Longident.t -> t ->
+  ((label_description * (unit -> unit)) list,
+   Location.t * t * lookup_error) result
+val lookup_all_labels_from_type:
+  ?use:bool -> loc:Location.t -> Path.t -> t ->
+  (label_description * (unit -> unit)) list
+
+val lookup_instance_variable:
+  ?use:bool -> loc:Location.t -> string -> t ->
+  Path.t * Asttypes.mutable_flag * string * type_expr
+
+(* Lookup by long identifiers. Raise [Not_found] if there would be an error. *)
+
+val find_value_by_name:
+  Longident.t -> t -> Path.t * value_description
+val find_type_by_name:
+  Longident.t -> t -> Path.t * type_declaration
+val find_module_by_name:
+  Longident.t -> t -> Path.t * module_declaration
+val find_modtype_by_name:
   Longident.t -> t -> Path.t * modtype_declaration
-val lookup_class:
-  ?loc:Location.t -> ?mark:bool ->
+val find_class_by_name:
   Longident.t -> t -> Path.t * class_declaration
-val lookup_cltype:
-  ?loc:Location.t -> ?mark:bool ->
+val find_cltype_by_name:
   Longident.t -> t -> Path.t * class_type_declaration
 
+val find_constructor_by_name:
+  Longident.t -> t -> constructor_description
+val find_label_by_name:
+  Longident.t -> t -> label_description
+
+(* Check if a name is bound *)
+
+val bound_value: string -> t -> bool
+val bound_module: string -> t -> bool
+val bound_type: string -> t -> bool
+val bound_modtype: string -> t -> bool
+val bound_class: string -> t -> bool
+val bound_cltype: string -> t -> bool
```

Comment:
FTR: it looks like only `bound_value` is ever used in the codebase. But I'd keep the functions here: giving compiler-libs a decent API seems good.